### PR TITLE
Seeding script

### PR DESCRIPTION
**Pull request by VJagiasi**
# Notion ticket link

[Create Seeding Script](https://www.notion.so/uwblueprintexecs/Create-Seeding-Script-84d4983a336e4db6b6dda03d8b77d4dc?pvs=4)

## Implementation description
- Added the seeding script using snaplet to implement seeding and used faker to generate date
- Added the seeding script running, sync and migration database commands to package.json
 

## Steps to test

It will automatically seed when you run `docker compose up --build`

### Steps to test manually*

1. Ensure the database is running locally and make sure you have initialized the database `npx prisma migrate dev --name init`
2. In the root directory,  `npx @snaplet/seed sync` to sync the seed file to any change in the data models of the database
3. `npx prisma db seed` to seed the database based on seed.ts
4. Run `docker exec -it sistema-db-1 /bin/bash` and then in the container run `psql -U sistema -d sistema` to enter the Postgres shell
6. In the Postgres shell check if the database tables are seeded by running `SELECT * FROM public."<Table-name>";` eg: `SELECT * FROM public."Absence";`

## What should reviewers focus on?
- All the tables are seeded correctly with correct data and are in sync with each other

## Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,32 +9,48 @@
...
+    "@prisma/client": "^5.15.1",
+    "@snaplet/copycat": "^5.0.0",
+    "@snaplet/seed": "^0.97.20",
     "framer-motion": "^6.3.0",
     "next": "^14.2.3",
     "next-auth": "^5.0.0-beta.19",
+    "postgres": "^3.4.4",
     "prisma": "^5.15.0",
     "react": "^18.2.0",
-    "react-calendar": "^5.0.0",
```
Comment (ChinemeremChigbo) [inline package.json:26]: Please re-add react-calendar and rerun `npm install` to fix the failing tests

**Comment by ChinemeremChigbo** [comment]
Please also add instructions for running the seeding script to the README.

**Comment by VJagiasi** [comment]
> Please also add instructions for running the seeding script to the README.

Added

**Review by ChinemeremChigbo — COMMENTED** [review]
Please fix your merge conflicts and make sure your seeding script matches the updated schema.

**Comment by VJagiasi** [comment]
@ChinemeremChigbo can you check the build error not sure if I can solve it

File: prisma/seed/seed.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      ]);
+      return {
+        name: subject,
+        abbreviation: generateSubjectAbbreviation(subject),
+      };
+    })
+  );
+
+  await seed.absence((createMany) =>
+    createMany(10, () => ({
```
Comment (phamkelly17) [inline prisma/seed/seed.ts:69]: doesn't this need to have a absentTeacher, subject, and location? It should have all the required fields here: https://www.notion.so/uwblueprintexecs/Data-Models-d5cdb66530834c8da4ace4184bb363f2#dd896001196c4e14a6e248ab7510af1f

**Review by ChinemeremChigbo — COMMENTED** [review]
Please fix the failing deployment.

File: README.md [hunk 3]
```diff
@@ -137,6 +137,69 @@ psql -U sistema -d sistema
...
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+```
+
+Run the following commands:
+
+```bash
+npx prisma db generate
```
Comment (ChinemeremChigbo) [inline README.md:156]: I get `! Unknown command "generate"`

File: README.md [hunk 4]
```diff
@@ -137,6 +137,69 @@ psql -U sistema -d sistema
...
+# check if the tables are seeded correctly
+SELECT * FROM public."<Table-name>";
+
+```
+
+Vercel Postgres: make sure you pull the new secrets from hashicorp.
+
+In schema.prisma, replace datasource with:
+
+```bash
```
Comment (ChinemeremChigbo) [inline README.md:176]: Similar text is repeated twice. I am assuming this is supposed to replace the text above

File: prisma/seed/seed.config.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { SeedPrisma } from '@snaplet/seed/adapter-prisma';
+import { defineConfig } from '@snaplet/seed/config';
+import { PrismaClient } from '@prisma/client';
+
+// export default defineConfig({
```
Comment (ChinemeremChigbo) [inline prisma/seed/seed.config.ts:5]: Please remove

**Review by ludavidca — APPROVED** [review]
Everything looks ready to merge in.

**Comment by ChinemeremChigbo** [comment]
> Conditional-Postinstall seems to be required for the seeding script to not interfere with deployment. I tried removing it and reverting to only using start.sh, but the deployments failed after that occurred, which is likely due to Prisma needing different Postinstall scripts to work locally and in deployment. Error messages could not be viewed due to a lack of Vercel pro accounts.

Fixed the failing builds. We need to sync @ludavidca @anantduggal @RahulTandon1 to fix the lack of build introspection.

File: Dockerfile [hunk 6]
```diff
@@ -6,12 +6,18 @@ WORKDIR /sistema
...
+RUN chmod +x start.sh
 
 # Expose the port the app runs on
 EXPOSE 3000
 
-# Start the Next.js application in development mode
-CMD ["npm", "run", "dev"]
\ No newline at end of file
+# Use the startup script as the entry point
+ENTRYPOINT ["/bin/sh", "./start.sh"]
```
Comment (ChinemeremChigbo) [inline Dockerfile:23]: Why can't we run the commands in the shell script directly in the Dockerfile?

File: README.md [hunk 7]
```diff
@@ -85,6 +85,7 @@ git push -f
 ```
 
 ### Docker Commands
+*Make sure that the NON POOLING URL is changed to db not localhost before running"
```
Comment (ChinemeremChigbo) [inline README.md:88]: Should add a section for seeding the prod DB

File: README.md [hunk 8]
```diff
@@ -126,6 +127,13 @@ docker system prune -a --volumes
 
 ### Accessing PostgreSQL Database
 
+Run in two lines:
+```bash
+docker exec -it sistema-db-1 psql -U sistema -d sistema
+SELECT * FROM public."Absence";
+```
+
+Broken into Parts
```
Comment (ChinemeremChigbo) [inline README.md:136]: What does "Broken into Parts" mean?

File: docker-compose.yml [hunk 9]
```diff
@@ -30,4 +30,4 @@ services:
       - ./db-init:/docker-entrypoint-initdb.d
 
 volumes:
-  postgres_data:
\ No newline at end of file
+  postgres_data:
```
Comment (ChinemeremChigbo) [inline docker-compose.yml:33]: Why do we add a \n?

File: prisma/migrations/20240623012346_init/migration.sql
Comment (ChinemeremChigbo) [inline prisma/migrations/20240623012346_init/migration.sql:1]: Why do we need these migrations?

File: prisma/migrations/20240712012504_init/migration.sql
Comment (ChinemeremChigbo) [inline prisma/migrations/20240712012504_init/migration.sql:1]: Why do we need these migrations?

File: Dockerfile (hunk 6 above)
Comment (ludavidca) [reply Dockerfile:23]: patched

File: README.md (hunk 7 above)
Comment (ludavidca) [reply README.md:88]: resolved.

File: README.md (hunk 8 above)
Comment (ludavidca) [reply README.md:136]: Separating the commands into different lines. Rewrote for clarification

File: docker-compose.yml (hunk 9 above)
Comment (ludavidca) [reply docker-compose.yml:33]: :0

File: prisma/migrations/20240623012346_init/migration.sql
Comment (ludavidca) [reply prisma/migrations/20240623012346_init/migration.sql:1]: I think this is something from the DB PR.

File: prisma/seed/seed.ts (hunk 2 above)
Comment (ludavidca) [reply prisma/seed/seed.ts:71]: Resolved. Should be working. We bing chilling.

File: prisma/migrations/20240623012346_init/migration.sql
Comment (ludavidca) [reply prisma/migrations/20240623012346_init/migration.sql:1]: These have been removed

**Review by ludavidca — APPROVED** [review]
Looks good to me. All issues are resolved.

**Review by ChinemeremChigbo — APPROVED** [review]
LBTM

**Pull request merged**